README.md

```diff
@@ -13,175 +13,355 @@ Compartmentalization of host behavior hooks for JS.
-  staticImports(): {specifier: string, exportNames}[];
-}
-interface SourceTextStaticModuleRecord extends StaticModuleRecord {
-  // no coerce
-  constructor(source: string);
+// Static module records are an opaque token representing the compilation
+// of a module that can be reused across multiple compartments.
+interface StaticModuleRecord {
+  // Static module records can be constructed from source.
+  // XS allows third-party module records and source descriptors to
+  // be precompiled as well.
+  constructor(source: string | { source: string } | ThirdPartyStaticModuleRecord);
+
+  // Static module records reflect their bindings for information only.
+  // Compartments use internal slots for the compiled code and bindings.
+  bindings: Array<Binding>;
 }
+
+// A ModuleDescriptor captures a static module record and per-compartment metadata.
+type ModuleDescriptor = 
+  // Using a string for a module descriptor indicates a static module record
+  // should be inherited from the parent compartment, but not the parent
+  // compartment's instance.
+  string
+  // A static module record descriptor produces a unique instance
+  // of a module in a compartment.
+  | {
+    record: StaticModuleRecord | ThirdPartyStaticModuleRecord,
+
+    // An optional alias.
+    // For example, in a Node.js package compartment, '.' may imply a memo entry
+    // for './index.js'.
+    // TODO this design does not yet account for the possibility of multiple
+    // aliases, which have not been necessary for a faithful imitation of Node.js.
+    // That might be better accounted for with {compartment?, specifier}
+    // module descriptors, below.
+    specifier?: string,
+
+    // Properties to copy to the `import.meta` of the resulting module instance.
+    meta?: Object,
+  }
+  // As a shorthand for constructing a StaticModuleRecord,
+  // a module descriptor can provide source.
+  | {
+    source: string,
+    specifier?: string,
+    meta?: Object,
+  }
+  // To refer to a module instance in another compartment, we need either to
+  // have a module descriptor that can refer to it, or have a `module` method and
+  // use module exports namespaces as opaque tokens to refer to modules in foreign
+  // compartments.
+  | {
+    // A compartment instance.
+    // We do not preclude the possibility that compartment is the same compartment,
+    // as that is possible to achieve in moduleMapHook and loadHook, albeit not
+    // in moduleMap.
+    // TODO Idea: it may be sufficient to make this compartment instance optional,
+    // defaulting to the current instance, in order to implement arbitrary
+    // numbers of aliases to a single module instance, from one or many other
+    // compartments, which would obviate the need for `Compartmnt.prototype.module`.
+    compartment: Compartment,
+    // The full specifier of the corresponding module instance in the other compartment.
+    specifier: string,
+  }
+  | {
+    namespace: ModuleExportsNamespace,
+  };
```

Comment:
We (@erights, @patrick-soquet, @phoddie, @mhofman, myself) haven’t settled on whether to continue allowing a bare module exports namespace (distinguished by brand check) as a kind of module descriptor. I’m still leaning toward not supporting the abbreviation.

We also did settle on allowing the `{ namespace }` module descriptor shape. The behavior for brand-check-passing module namespace objects would be special and allow lazy binding. The behavior for any other object would be to capture a snapshot of the object with a get own property names check and a get for every property, such that it realizes a module namespace object that exports a binding for all those properties with those constant values.

Reply:
When I proposed the descriptor, I meant it to be instead of allowing the namespace directly. One of the goals was to preserve practical namespace transparency, without forcing existing membrane code to make a weird special case for module namespaces.

So do we agree that the descriptor is *instead* rather than *in addition to*?

MarkM: y

Reply:
Where you say "lazy binding" above, did you mean "live binding"?

Reply:
Yes, I mean live binding.

I would like to support:

```js
new Compartment({
  moduleMap: {
    x: {
      namespace: {}
    }
  }
})
```

And also support:

```js
import * as x from 'x';
new Compartment({
  moduleMap: {
    x: {
      namespace: x
    }
  }
})
```

And I would prefer not to support (even though I agree it is convenient):

```js
import * as x from 'x';
new Compartment({
  moduleMap: { x }
})
```

Reply:
Yes. I meant only to prohibit the last.